core/src/env2/buffer_arena.rs

```diff
@@ -0,0 +1,358 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This file defines the global buffer arena that is accessible globally
+//! and acts as a cache for allocated heap memory to avoid heap memory
+//! throttling resulting from heap usage for intermediate computations.
+//!
+//! Exactly this happens a lot in the boundaries between SRML contracts
+//! and ink! since encoding and decoding of SCALE values has to be done
+//! in such an intermediate buffer.
+//!
+//! Users and systems are advised to share a common set of allocated buffers
+//! provided by the global buffer arena.
+
+use crate::{
+    env2::utils::{
+        EnlargeTo,
+        Reset,
+    },
+    memory::vec::Vec,
+};
+use cfg_if::cfg_if;
+use core::cell::{
+    Cell,
+    RefCell,
+};
+
+/// The maximum amount of used byte buffers at the same time.
+///
+/// Since the whole point behind this byte buffer arena is to cache
+/// allocated heap memory the number of concurrent byte buffers that are
+/// in use at the same time should be kept small.
+const IN_USE_LIMIT: usize = 1000;
```

Comment:
So the limit is defined by the number of instances, but what is the byte limit per instance? I wonder if it makes more sense to define the arena limit as a number of bytes, since this would provide a better idea of the upper byte boundary.

Does the number have any reasoning behind it or is it more a rough value of thumb for now?

Reply:
The limit is arbitrary.
However imposing it makes sense since there is one way how you could abuse this global buffer arena and it is when you end up with many many different buffers, so when `in_use` is a high number.
I expect normal use cases to spawn at most 5-10 buffers so 1000 is already pretty intense.

A regulation on the total bytes in use could be introduced but it should definitely be possible to have for example one giant buffer if that's what has been needed.